File: src/com/nimbusly/isometrix/GameRenderer.java
package com.nimbusly.isometrix;

import java.io.IOException;
import java.io.InputStream;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.FloatBuffer;

import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.opengl.GLSurfaceView;
import android.opengl.GLU;
import android.opengl.GLUtils;
import android.opengl.GLSurfaceView.Renderer;
import android.util.Log;

public class GameRenderer implements Renderer {

	protected Context context;
	protected int screenWidth, screenHeight;

	private static final String TAG = "GameRenderer";
	
	protected long startTime;
	protected long frameCount = 0;
	StringBuffer logMessage = new StringBuffer(100);
	
	Boolean stopped;

	public GameRenderer(Context context) {
		this.context = context; 
	}

	/**
	 * Called from another thread to stop rendering thread
	 * @param view
	 */
	public void stopRendering(final GLSurfaceView view) {
		view.queueEvent(new Runnable() {
			public void run() {
	            GameRenderer.this.stop(view);
	        }
		});
        synchronized (stopped) {
            while (!stopped) {
                try { stopped.wait(); } catch (InterruptedException e) {}
            }
        }
	}

	/**
	 * This supports a work around to a race condition in pause. To pause renderer from input thread,
	 * queueEvent(
	 * @param view
	 */
	public void stop(GLSurfaceView view) {
		view.onPause();
		synchronized (stopped) {
			stopped = true;
		}
	}
	/**
	 * Initialize the GL surface
	 */
	public void onSurfaceCreated(GL10 gl, EGLConfig config) {
		gl.glEnable(GL10.GL_TEXTURE_2D);			//Enable Texture Mapping ( NEW )
		gl.glShadeModel(GL10.GL_FLAT); 			//Enable Smooth Shading
		gl.glClearColor(0.5f, 0.5f, 0.5f, 1.0f); 	//Black Background
		gl.glDisable(GL10.GL_DEPTH_TEST); 			//Enables Depth Testing
        gl.glDisable(GL10.GL_DITHER);
        gl.glDisable(GL10.GL_LIGHTING);
		gl.glHint(GL10.GL_PERSPECTIVE_CORRECTION_HINT, GL10.GL_FASTEST);
		startTime = System.currentTimeMillis();
	}

	/**
	 * If the surface changes, reset the view
	 */
	public void onSurfaceChanged(GL10 gl, int width, int height) {
		if(height == 0) { 						//Prevent A Divide By Zero By
			height = 1; 						//Making Height Equal One
		}
		screenWidth = width;
		screenHeight = height;

		gl.glViewport(0, 0, width, height); 	//Reset The Current Viewport
        gl.glMatrixMode(GL10.GL_PROJECTION); 	//Select The Projection Matrix
		gl.glOrthof(0.0f, width, 0.0f, height, 0.0f, 1.0f);
		gl.glLoadIdentity(); 					//Reset The Projection Matrix

		gl.glMatrixMode(GL10.GL_MODELVIEW); 	//Select The Modelview Matrix
		gl.glLoadIdentity(); 					//Reset The Modelview Matrix
	}
	
	/**
	 * Here we do our drawing
	 */
	public void onDrawFrame(GL10 gl) {
		long currentTime = System.currentTimeMillis();
		long elapsed = currentTime - startTime;
		if (currentTime > startTime + 10000) {
			reportFramesPerSecond(elapsed, frameCount);
			startTime = currentTime;
			frameCount = 0;
		}
		frameCount += 1; 
	}

	public void reportFramesPerSecond(long elapsed, long frameCount) {
		Log.d(TAG, "1 frame took " + (elapsed / frameCount) + "ms; FPS = " + (frameCount * 1000 / elapsed));
	}
	
}

File: src/com/nimbusly/isometrix/GameRenderer.scala
package com.nimbusly.isometrix

import javax.microedition.khronos.egl.EGLConfig;
import javax.microedition.khronos.opengles.GL10;

import android.content.Context;
import android.opengl.GLSurfaceView;
import android.opengl.GLSurfaceView.Renderer;
import android.util.Log;

/**
 * GameRenderer is a custom GL Renderer that is responsible for rendering the game 
 * into the GLSurfaceView. Rendering begins when this class is added to a GlSurfaceView,
 * although it can be stopped and restarted as needed.
 * 
 * @author semprebon
 *
 */
class GameRenderer(context: Context) extends Renderer {
	protected var screenWidth = 1;
	protected var screenHeight = 1;

	private val TAG = "GameRenderer"
	
	protected var startTime = 0L;
	protected var frameCount = 0;
	protected var logMessage: StringBuffer = new StringBuffer(100);
	
	var stopped: Boolean = true;
	var lock: AnyRef = new Object()

	implicit def runnable(f: () => Unit): Runnable =
		new Runnable() { def run() = f() }

	/**
	 * Called from another thread to stop rendering thread.
	 * 
	 * stopRendering first queues an even to set a stop flag, then waits for that
	 * event to be executed and the flag to be set.
	 * to
	 * 
	 * @param view the GLSurfaceView this renderer is rendering to
	 */
	def stopRendering(view: GLSurfaceView): Unit = {
		view.queueEvent(runnable(() => GameRenderer.this.stop(view)))
        lock.synchronized {
            while (!stopped) {
               	lock.wait(); 
            }
        }
	}

	/**
	 * Stops the view rendering from rendering thread
	 *
	 * @param view the GLSurfaceView this renderer is rendering to
	 */
	def stop(view: GLSurfaceView) {
		view.onPause();
		lock.synchronized {
			stopped = true;
		}
	}
	
	/**
	 * Initialize the GL surface.
	 */
	def onSurfaceCreated(gl: GL10, config: EGLConfig) {
		gl.glEnable(GL10.GL_TEXTURE_2D);			//Enable Texture Mapping ( NEW )
		gl.glShadeModel(GL10.GL_FLAT); 			//Enable Smooth Shading
		gl.glClearColor(0.5f, 0.5f, 0.5f, 1.0f); 	//Black Background
		gl.glDisable(GL10.GL_DEPTH_TEST); 			//Enables Depth Testing
        gl.glDisable(GL10.GL_DITHER);
        gl.glDisable(GL10.GL_LIGHTING);
		gl.glHint(GL10.GL_PERSPECTIVE_CORRECTION_HINT, GL10.GL_FASTEST);
		Log.d(TAG, "Extensions:" + gl.glGetString(GL10.GL_EXTENSIONS))
		Log.d(TAG, "Version:" + gl.glGetString(GL10.GL_VERSION))
		startTime = System.currentTimeMillis();
	}

	/**
	 * Reset anything that needs to be reset if the surface changes (as in dimension changes)
	 */
	def onSurfaceChanged(gl: GL10, width : Int, height : Int) {
		
		if (height == 0) { 						//Prevent A Divide By Zero By
			return 						//Making Height Equal One
		}
		screenWidth = width;
		screenHeight = height;
		Log.d(TAG, "Setting window size to " + width + "x" + height)
		gl.glViewport(0, 0, width, height); 	//Reset The Current Viewport
        gl.glMatrixMode(GL10.GL_PROJECTION); 	//Select The Projection Matrix
		gl.glOrthof(0.0f, width, 0.0f, height, 0.001f, 100.0f);
		//gl.glLoadIdentity(); 					//Reset The Projection Matrix

		gl.glMatrixMode(GL10.GL_MODELVIEW); 	//Select The Modelview Matrix
		gl.glLoadIdentity(); 					//Reset The Modelview Matrix
	}
	
	/**
	 * Render the scene. You should override this class to implement your game-specific drawing.
	 * 
	 * TODO: this should probably call an abstract method to do the drawing
	 */
	def onDrawFrame(gl: GL10) {
		val currentTime = System.currentTimeMillis();
		val elapsed = currentTime - startTime;
		if (currentTime > startTime + 10000) {
			reportFramesPerSecond(elapsed, frameCount);
			startTime = currentTime;
			frameCount = 0;
		}
		frameCount += 1; 
	}

	/**
	 * Report fromes per second to the log (allocates memory)
	 * 
	 * @param elapsed
	 * @param frameCount
	 */
	def reportFramesPerSecond(elapsed : Long, frameCount : Long) {
		Log.d(TAG, "1 frame took " + (elapsed / frameCount) + "ms; FPS = " + (frameCount * 1000 / elapsed))
	}
	
}